src/H5Dcompact.c

```diff
@@ -47,6 +47,24 @@
 /* Local Typedefs */
 /******************/
 
+/* Callback info for readvv operation when file driver
+ * uses special memory management functions
+ */
+typedef struct H5D_compact_readvv_special_mem_ud_t {
+    H5F_shared_t *f_sh;   /* Shared file for dataset */
+    void *        rbuf;   /* Pointer to buffer to be read from */
+    void *        dstbuf; /* Pointer to buffer to be read into */
+} H5D_compact_readvv_special_mem_ud_t;
+
+/* Callback info for writevv operation when file driver
+ * uses special memory management functions
+ */
+typedef struct H5D_compact_writevv_special_mem_ud_t {
+    H5F_shared_t *f_sh;   /* Shared file for dataset */
+    const void *  wbuf;   /* Pointer to buffer to be written from */
+    void *        dstbuf; /* Pointer to buffer to be written into */
+} H5D_compact_writevv_special_mem_ud_t;
```

Comment:
While annoying, use two different structures similar to the approach in H5Dcontig.c for const-correctness without warnings.